File: 2025/02 February/22 Longest valid Parentheses/main.cpp
// Link: https://www.geeksforgeeks.org/problems/longest-valid-parentheses5657/1

#include <iostream>
#include <stack>
using namespace std;

class Solution
{
public:
    int maxLength(string &s)
    {
        stack<int> st;
        st.push(-1);
        int max_length = 0;

        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == '(')
            {
                st.push(i);
            }
            else
            {
                st.pop();
                if (st.empty())
                {
                    st.push(i);
                }
                else
                {
                    max_length = max(max_length, i - st.top());
                }
            }
        }

        return max_length;
    }
};

int main()
{
    Solution object;
    string s = "((()";
    cout << object.maxLength(s) << endl;

    s = ")()())";
    cout << object.maxLength(s) << endl;

    s = "())()";
    cout << object.maxLength(s) << endl;

    return 0;
}

File: 2025/02 February/22 Longest valid Parentheses/main.py
# Link: https://www.geeksforgeeks.org/problems/longest-valid-parentheses5657/1


class Solution:
    def maxLength(self, s):
        stack = [-1]
        max_length = 0
        for i in range(len(s)):
            if s[i] == "(":
                stack.append(i)
            else:
                stack.pop()
                if len(stack) == 0:
                    stack.append(i)
                else:
                    max_length = max(max_length, i - stack[-1])
        return max_length


if __name__ == "__main__":
    obj = Solution()
    s = "((()"
    print(obj.maxLength(s))

    s = ")()())"
    print(obj.maxLength(s))

    s = "())()"
    print(obj.maxLength(s))
